src/jobs/keys-update/keys-update.service.ts

```diff
@@ -133,51 +143,138 @@ export class KeysUpdateService {
           const prevNonce = moduleInStorage?.nonce;
-          // update staking module information
-          await this.srModulesStorage.upsert(contractModule, currNonce);
 
           this.logger.log(`Nonce previous value: ${prevNonce}, nonce current value: ${currNonce}`);
 
-          if (prevNonce === currNonce) {
-            this.logger.log("Nonce wasn't changed, no need to update keys");
-            // case when prevELMeta is undefined but prevNonce === currNonce looks like invalid
-            // use here prevElMeta.blockNumber + 1 because operators were updated in database for  prevElMeta.blockNumber block
-            if (
-              prevElMeta &&
-              prevElMeta.blockNumber < currElMeta.number &&
-              (await moduleInstance.operatorsWereChanged(
-                contractModule.stakingModuleAddress,
-                prevElMeta.blockNumber + 1,
-                currElMeta.number,
-              ))
-            ) {
-              this.logger.log('Update events happened, need to update operators');
-              await moduleInstance.updateOperators(contractModule.stakingModuleAddress, currElMeta.hash);
-            }
+          if (!prevElMeta) {
+            this.logger.log('No past state found, start indexing', { stakingModuleAddress, currentBlockHash });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
+            continue;
+          }
+
+          if (prevNonce !== currNonce) {
+            this.logger.log('Nonce has been changed, start indexing', {
+              stakingModuleAddress,
+              currentBlockHash,
+              prevNonce,
+              currNonce,
+            });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
+            continue;
+          }
 
+          if (this.isTooMuchDiffBetweenBlocks(prevElMeta.blockNumber, currElMeta.number)) {
+            this.logger.log('Too much difference between the blocks, start indexing', {
+              stakingModuleAddress,
+              currentBlockHash,
+            });
+
+            await moduleInstance.update(stakingModuleAddress, currentBlockHash);
+            await this.srModulesStorage.upsert(contractModule, currNonce, currentBlockHash);
+            lastChangedBlockHash = currentBlockHash;
             continue;
           }
 
-          await moduleInstance.update(contractModule.stakingModuleAddress, currElMeta.hash);
+          // calculate once per iteration
+          // no need to recheck each module separately
+          isReorgDetected = isReorgDetected ? true : await this.isReorgDetected(prevElMeta.blockHash, currentBlockHash);
```

Comment:
Do i understand right that for every staking module we check reorg ? But prevElMeta.blockHash currentBlockHash 
we get from common tabl and before cycle 'for'. Do we need to put isReorgDetected detection outside of 'for' cycle ?

Reply:
We only check the reorganization if it is necessary. We check once and cache, the function itself is called once for all modules.

Reply:
understood